rfd/0159-password-status.md

```diff
@@ -0,0 +1,137 @@
+
+Rationale: as there is no reliable way to verify that the user actually has a password configured, we _have to_ allow setting a password without providing the current password for those whose password state is unknown (unspecified). We could potentially require current password for those users whose password state is _known_ (see [Recognizing Users Who Configured Their Passwords](#recognizing-users-who-configured-their-passwords)). This, however, is not necessary: if a potential attacker is able to prove their identity using a passwordless key (as opposed to an MFA key or an app token), they already are able to sign in and perform _any_ other action, including adding their own passwordless keys, as well as removing existing passwordless keys of the user under attack. 
+
+The user will need to pick upfront whether they want to authenticate with a passwordless or an MFA device, even in case of hardware tokens. This is because in order to perform a WebAuthn authentication ceremony, we need to specify upfront whether we require [user verification](https://w3c.github.io/webauthn/#user-verification) or not using the [`PublicKeyCredentialRequestOptions.userVerification`](https://w3c.github.io/webauthn/#dom-publickeycredentialrequestoptions-userverification) field. The client (user agent) is not obliged to return an assertion that contains user verification if we didn't require it, and although the WebAuthn spec says it should do it ["if possible"](https://w3c.github.io/webauthn/#dom-userverificationrequirement-preferred), the exact details of what is considered "possible" are left to interpretation. The key question of whether we are guaranteed to receive user verification if a token capable of user verification has been used is left unanswered. Since user verification is required for changing password without giving the current one, we can't allow the user to perform it if the returned authenticator data contains flags with the [User Verified bit](https://w3c.github.io/webauthn/#authdata-flags-uv) set to 0. Thus, we need to specify user verification requirement when we create the authentication request.
+
+The following diagram sums up the modified process:
+
+```mermaid
+flowchart TD
+Start([Start]) --> MFAAvail{MFA or passkeys<br>available?}
+MFAAvail --> |Yes| ShowMFAOptions[Show available identity<br>confirmation options]
+ShowMFAOptions --> Pick{Which<br>option?}
+Pick --> |Passwordless| WebAuthnVer[WebAuthn authentication,<br>verification required]
+Pick --> |MFA token| WebAuthnNoVer[WebAuthn authentication,<br>verification preferred]
```

Comment:
Unless you plan to actually modify challenge issuance we will default to "discouraged" for "MFA" checks: https://github.com/gravitational/teleport/blob/5c636836458ab562ee350384553eabdde7081065/lib/auth/webauthn/config.go#L50-L53

Note that this is alright because the password is doing the "verification".

```suggestion
Pick --> |MFA token| WebAuthnNoVer[WebAuthn authentication,<br>verification discouraged]
```